src/main/kotlin/org/ageseries/libage/data/Quantity.kt

```diff
@@ -0,0 +1,207 @@
+interface ReciprocalArealDensity
+interface Density
+interface Substance
+interface Area
+interface Volume
+interface Temp // Rename once we remove Temperature
+interface MolarConcentration
+interface SpecificHeatCapacity
+interface HeatCapacity
+interface ThermalConductivity
+interface Pressure
+
+/**
+ * Represents a physical quantity, characterised by a [Unit] and a real number [value].
+ * */
+@JvmInline
+value class Quantity<Unit>(val value: Double) : Comparable<Quantity<Unit>> {
+    constructor(quantity: Double, s: QuantityScale<Unit>) : this(s.scale.unmap(quantity))
+
+    val isZero get() = value == 0.0
+
+    /**
+     * Gets the numerical value of this quantity.
+     * */
+    operator fun not() = value
+
+    operator fun unaryMinus() = Quantity<Unit>(-value)
+    operator fun unaryPlus() = Quantity<Unit>(+value)
+    operator fun plus(b: Quantity<Unit>) = Quantity<Unit>(this.value + b.value)
+    operator fun minus(b: Quantity<Unit>) = Quantity<Unit>(this.value - b.value)
+    operator fun times(scalar: Double) = Quantity<Unit>(this.value * scalar)
+    operator fun div(scalar: Double) = Quantity<Unit>(this.value / scalar)
+    /**
+     * Divides the quantity by another quantity of the same unit. This, in turn, cancels out the quantity, returning the resulting number.
+     * */
+    operator fun div(b: Quantity<Unit>) = this.value / b.value
+
+    override operator fun compareTo(other: Quantity<Unit>) = value.compareTo(other.value)
+
+    operator fun compareTo(b: Double) = value.compareTo(b)
+
+    /**
+     * Maps this quantity to another scale of the same unit.
+     * */
+    operator fun rangeTo(s: QuantityScale<Unit>) = s.scale.map(value)
+
+    override fun toString() = value.toString()
+
+    fun <U2> reparam(factor: Double = 1.0) = Quantity<U2>(value * factor)
+}
+
+fun <U> min(a: Quantity<U>, b: Quantity<U>) = Quantity<U>(kotlin.math.min(!a, !b))
+fun <U> max(a: Quantity<U>, b: Quantity<U>) = Quantity<U>(kotlin.math.max(!a, !b))
+fun <U> abs(q: Quantity<U>) = Quantity<U>(kotlin.math.abs(!q))
+
+/**
+ * Defines the standard scale of the [Unit] (a scale with factor 1).
+ * */
+fun <Unit> standardScale(factor: Double = 1.0, base: Double = 0.0) = QuantityScale<Unit>(factor, base)
+
+val KILOGRAMS = standardScale<Mass>()
+val GRAMS = -KILOGRAMS
+
+val SECOND = standardScale<Time>()
+val MILLISECONDS = -SECOND
+val MICROSECONDS = -MILLISECONDS
+val NANOSECONDS = -MICROSECONDS
+val MINUTES = SECOND * 60.0
+val HOURS = MINUTES * 60.0
+val DAYS = HOURS * 24.0
+
+val METER = standardScale<Distance>()
+val CENTIMETERS = METER / 100.0
+val MILLIMETERS = -METER
+
+val JOULE = standardScale<Energy>()
+val KILOJOULES = +JOULE
+val MEGAJOULES = +KILOJOULES
+val GIGAJOULES = +MEGAJOULES
+val WATT_SECONDS = QuantityScale<Energy>(JOULE.factor, 0.0)
+val WATT_MINUTES = WATT_SECONDS * 60.0
+val WATT_HOURS = WATT_MINUTES * 60.0
+val KW_HOURS = WATT_HOURS * 1000.0
+
+// Serious precision issues? Hope not! :Fish_Smug:
```

Comment:
You're so lucky GitHub doesn't have my numerous dragon emoji _yet_

![fisheyes](https://cdn.discordapp.com/emojis/697887785389916211.webp?size=240&quality=lossless)

Reply:
Yet.....